src/comm_droplet.F

```diff
@@ -660,23 +660,190 @@ end subroutine setupIndexPointers
 
       use input, only : mynw,mysw,myne,myse,myeast,mywest,mynorth,mysouth, &
-        nparcelsLocalActive
+        nparcelsLocal,nparcelsLocalActive,myid
       use constants, only: num_nn,inorth,isouth, &
                          iwest,ieast,inw,ine,isw,ise,undefined_index
+#ifdef _OPENACC
+      use cutensorex
+#endif
 
       integer, intent(inout) :: ptrDepart(num_nn)
+      integer, intent(in) :: Depart(num_nn)
 
       integer, intent(inout) :: Depart_ind(:)
       integer, intent(in) :: pdata_neighbor(:)
 
       ! local variables
       integer :: n_idx,s_idx,e_idx,w_idx,nw_idx,ne_idx,sw_idx,se_idx
       integer np
+      integer :: Cidx(nparcelsLocal)
+
+      integer :: numDepart,numerrors
+      logical, parameter :: verbose = .false.
+      !$acc declare present(Depart_ind,pdata_neighbor)
 
 
+      numDepart=SUM(Depart)
+
+      n_idx  = ptrDepart(inorth)
+      s_idx  = ptrDepart(isouth)
+      w_idx  = ptrDepart(iwest)
+      e_idx  = ptrDepart(ieast)
+      nw_idx = ptrDepart(inw)
+      ne_idx = ptrDepart(ine)
+      sw_idx = ptrDepart(isw)
+      se_idx = ptrDepart(ise)
+      if(verbose .and. (numDepart.gt.0)) then
+         write(*,110) myid,'Depart:',Depart
+         write(*,110) myid,'ptrDepart:',ptrDepart
+      endif
+
+#if (defined(_OPENACC) && defined(_COUNTPREFIX))
+
+!$acc data create(Cidx)
+
+      if(Depart(inorth).gt.0) then
+
+         !JMD don't know why I get the following error when calling this
+         !function: DEALLOCATE: an illegal memory access was encountered
+         !call GatherIndices(inorth,ptrDepart,Cidx,pdata_neighbor,Depart_ind)
+         Cidx  = count_prefix(mask=(pdata_neighbor .eq. inorth),exclusive=.true.)
+         !$acc parallel loop gang vector default(present)
+         do np=1,nparcelsLocalActive-1
+            if(Cidx(np) .ne. Cidx(np+1)) then
+               Depart_ind(n_idx+Cidx(np))=np
+            endif
+         enddo
+         ! special treament for the last value in the array
+         !$acc parallel default(present) num_gangs(1) num_workers(1)
+         if(pdata_neighbor(nparcelsLocalActive).eq.inorth) then
+            Depart_ind(n_idx+Cidx(nparcelsLocalActive)) = nparcelsLocalActive
+         endif
+         !$acc end parallel
+      endif
+
+      if(Depart(isouth).gt.0) then
+         Cidx  = count_prefix(mask=(pdata_neighbor .eq. isouth),exclusive=.true.)
+         !$acc parallel loop gang vector default(present)
+         do np=1,nparcelsLocalActive-1
+            if(Cidx(np) .ne. Cidx(np+1)) then
+               Depart_ind(s_idx+Cidx(np))=np
+            endif
+         enddo
+         ! special treament for the last value in the array
+         !$acc parallel default(present) num_gangs(1) num_workers(1)
+         if(pdata_neighbor(nparcelsLocalActive).eq.isouth) then
+            Depart_ind(s_idx+Cidx(nparcelsLocalActive)) = nparcelsLocalActive
+         endif
+         !$acc end parallel
```

Comment:
This code structure has been repeated eight times here. Can we write a subroutine interface instead to reduce the potential code bug due to copy & paste?

Reply:
@sjsprecious  I actually did write a subroutine interface to execute these calculations, it is called GatherIndices.  Unfortunately, it suffers from a runtime error. I would appreciate it if you could take a look at it and see if you can see why it is not working.

Reply:
Ah, I see. That is why the subroutine is defined but never called in this PR. I will take a look at it.

Reply:
@johnmauff I have revised the code to make the `GatherIndices` subroutine work, at the cost of creating `Cidx` on the GPU every time this subroutine is called. I am not sure how large the impact is on the GPU performance.

However, this revision does not seem necessary to me. Since `count_prefix` is introduced in `nvhpc/23.5` (https://docs.nvidia.com/hpc-sdk/compilers/cuda-fortran-prog-guide/#cfref-fort-mods-cutensorex-cpfx), I suspect that this is just another compiler bug. Feel free to revert my commit if you do not like this change.